File: common/src/main/java/generations/gg/generations/core/generationscore/world/dialogue/DialogueNodeType.java
package generations.gg.generations.core.generationscore.world.dialogue;

import com.mojang.serialization.Codec;
import com.pokemod.pokemod.world.npc.dialogue.nodes.AbstractNode;
import generations.gg.generations.core.generationscore.world.dialogue.nodes.AbstractNode;

public record DialogueNodeType<T extends AbstractNode>(Codec<T> codec, Class<? extends AbstractNode> nodeClass) {
}

File: common/src/main/java/generations/gg/generations/core/generationscore/world/dialogue/DialogueNodeType.kt
package generations.gg.generations.core.generationscore.world.dialogue

import com.mojang.serialization.Codec
import generations.gg.generations.core.generationscore.world.dialogue.nodes.AbstractNode
import net.minecraft.network.FriendlyByteBuf

data class DialogueNodeType<T : AbstractNode>(
    @JvmField val codec: Codec<T>, val nodeClass: Class<out AbstractNode>, private val decoder: (FriendlyByteBuf) -> T) {
    fun decode(buf: FriendlyByteBuf): AbstractNode = decoder.invoke(buf);
}
